crypto/stark-verifier-ethereum/contracts/primefield.sol

```diff
@@ -0,0 +1,124 @@
+pragma solidity ^0.6.4;
+
+contract PrimeField {
+    uint256 constant internal K_MODULUS = 0x800000000000011000000000000000000000000000000000000000000000001;
+    uint256 constant internal K_MODULUS_MASK = 0x0fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff;
+    uint256 constant internal K_MONTGOMERY_R = 0x7fffffffffffdf0ffffffffffffffffffffffffffffffffffffffffffffffe1;
+    uint256 constant internal K_MONTGOMERY_R_INV = 0x40000000000001100000000000012100000000000000000000000000000000;
+    uint256 constant internal GENERATOR = 3;
+    uint256 constant internal ONE_VAL = 1;
+    uint256 constant internal gen1024_val = 0x659d83946a03edd72406af6711825f5653d9e35dc125289a206c054ec89c4f1;
+
+    function from_montgomery(uint256 val) internal pure returns (uint256 res) {
+        assembly {
+            res := mulmod(val, K_MONTGOMERY_R_INV,
+                K_MODULUS)
+        }
+        return res;
+    }
+
+    function from_montgomery_bytes(bytes32 bs) internal pure returns (uint256) {
+        /// Assuming bs is a 256bit bytes object, in montgomery form, it is read into a field
+        /// element
+        uint256 res = uint256(bs);
+        return from_montgomery(res);
+    }
+
+    // This is an unchecked cast and should be used very carefully,
+    // and only in cases when the data is already in the right form.
+    function from_bytes_raw(bytes32 bs) internal pure returns(uint256 data) {
+        assembly {
+            data := bs
+        }
+    }
+
+    // This is an unchecked cast and should be used very carefully,
+    // and only in cases when the data is already in the right form.
+    function from_bytes_array_raw(bytes32[] memory input) internal pure returns(uint256[] memory data) {
+        assembly {
+            data := input
+        }
+    }
+
+    function to_montgomery_int(uint256 val) internal pure returns (uint256 res) {
+        assembly {
```

Comment:
No need for assembly here and in following functions: https://solidity.readthedocs.io/en/v0.6.4/units-and-global-variables.html#mathematical-and-cryptographic-functions